### ISSUE TITLE
Unexpected error when wrapping `self` with `ref`

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#6335]
### Discussed in https://github.com/microsoft/pyright/discussions/6334

<div type='discussions-op-text'>

<sup>Originally posted by **rchl** November  5, 2023</sup>
I don't understand why pyright is showing an error here:

```py
import weakref


class Foo:
    def __init__(self, foo: 'weakref.ref[Bar]'):
        pass


class Bar:
    def test(self) -> None:
        foo = Foo(weakref.ref(self))
```

Error

```
/usr/local/test.py
  /usr/local/test.py:11:19 - error: Argument of type "ref" cannot be assigned to parameter "foo" of type "ref" in function "__init__"
    "_weakref.ReferenceType" is incompatible with "_weakref.ReferenceType"
      Type parameter "_T@ReferenceType" is invariant, but "Self@Bar" is not the same as "Bar" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

mypy doesn't raise any issues and I believe that pyright also didn't in some earlier versions.</div>

**Comment by erictraut** [comment]
I agree this is a bug. It appears that it's related to bidirectional type inference for constructor calls when the `__new__` method is annotated to return `Self`. 

```python
from typing import Generic, Self, TypeVar

T = TypeVar("T")

class A(Generic[T]):
    def __new__(cls, o: T) -> Self:
        ...

a: A[float] = A(0) # False positive error
```

Interestingly, the problem doesn't occur when using the new PEP 695 syntax:

```python
from typing import Self

class B[T]:
    def __new__(cls, o: T) -> Self:
        ...

b: B[float] = B(0) # No Error
```

I'll create a bug report out of this discussion thread.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.335, which I just published. It will also be included in a future release of pylance.
